include/mata/synchronized_iterator.hh

```diff
@@ -0,0 +1,260 @@
+
+            std::vector<Iterator> positions;
+            std::vector<Iterator> ends;
+
+            explicit SynchronizedIterator(const int size = 0)
+            {
+                positions.reserve(size);
+                ends.reserve(size);
+            };
+
+            // This is supposed to be called only before an iteration,
+            // after constructor of reset.
+            // Calling after advance breaks the iterator.
+            virtual void push_back (const OrdVector<Key> &vector) {
+                // hope that the parameter is not deep copied ... is it?
+                this->positions.push_back(vector.begin());
+                this->ends.push_back(vector.end());
+            };
+
+            // Empties positions and ends.
+            // Though they should keep the allocated space.
+            void reset(const int size = 0) {
+                positions.clear();
+                ends.clear();
+                if (size > 0) {
+                    positions.reserve(size);
+                    ends.reserve(size);
+                }
+            };
+
+            virtual bool advance() = 0;
+            virtual const std::vector<Iterator> get_current() = 0;
+        };
+
+
+
+        template<typename Key>
+        class SynchronizedUniverzalIterator: public SynchronizedIterator<Key> {
+        public:
+
+            using Iterator = typename OrdVector<Key>::const_iterator;
+
+            // "minimum" would be the smallest class bounded from below by all positions that appears in all OrdVectos.
+            // Are all positions at this class?
+            // Invariant: it can be true only if all positions are indeed synchronized.
+            bool synchronized_at_current_minimum = false;
+
+            // advance() advances all positions to the NEXT minimum and returns true,
+            // or returns false if positions cannot be synchronized.
+            // If positions are synchronized to start with,
+            // then synchronized_at_current_minimum decides whether to stay or advance further.
+            // The general of the algorithm is to synchronize everybody with position[0].
+            bool advance()
+            {
+                //Nothing to synchronize.
+                if (this->positions.empty())
+                    return false; //?? or not?
+
+                // If already synchronized, start moving forward by advancing position[0] (and so break synchronization).
+                if (this->synchronized_at_current_minimum) {
+                    this->positions[0]++;
+                    synchronized_at_current_minimum = false;
+                }
+
+                // If positions[0] has nowhere to go, then sync is not possible.
+                if (this->positions[0] == this->ends[0])
+                    return false;
+
+                // Synchronise all positions with position[0].
+                for (size_t i = 1; i < this->positions.size(); i++)
```

Comment:
My usual comment (thanks Martin for letting me this): extract call to `size` outside of the loop. If by any chance, teh call to `size()` becomes linear, then the loop becomes quadratic.

Also switch `i++` to `++i`, which is usually more efficient.

I know these are small things, but we want iterators as much efficient as possible.

Reply:
ok, did not know about i++ ++i, why not.
size() - just spend half an hour debuging this, the size changes in one of the for loops, argh. Does it make sense to do these things, even though it makes the code less straightfoeward and more bug prone? Doesn't compiler do it anyway?

Reply:
If the size() changes during the call, then, I think it is OK, but it is a little bit shady. Will it increase or decrease? Is it safe? Maybe it is a bug actually :). I suggest to extract `size()` calls only if it stays constant during the for (sorry, didn't think of the case when the size() changes during the loop.

Reply:
As for the "compiler does it", I would be careful and not assume, that it does all kinds of optimizations (mostly it does; but in these two cases I'm not sure; it does a lot of inlining, e.g., but here it would require more complex analysis imo) and that all compilers do it.

Reply:
whatever, it is now finished and seeled forewer.